npg/man/cat1/xray.1



User Commands                                             XRAY(1)



NAME
     xray - Create SXR files based on skull xray tracings and PET
     image files. It is the Unix version of _s_a_r_p_l_o_c_s.


SYNOPSIS
     xray _s_x_r_f_i_l_e [_r_a_s_f_i_l_e _i_m_g_f_i_l_e _t_h_r_e_s_h_o_l_d _p_e_a_k_t_y_p_e _t_r_i_m]


THE REFERENCE SLICE
     The  default  reference  slice  is  the  lead   line.   This
     corresponds to slice 16 if the offset is zero.

     The default reference slice used is slice 16.  However,  the
     user  can  change  the reference slice in the ``Parameters''
     menu. To make sure that the reference slice falls inside the
     brain, the following heuristic algorithm is used:

     1) Find all the slices that have the maximum length  on  the
     skull x ray inside the brain. The default reference slice is
     used to compute the lengths.

     2) Pick the center slice which has the maximum length as the
     new reference slice. If the center slice is greater than 28,
     use slice 28 as the new reference slice. On the other  hand,
     if  the  center slice is less than 4, slice 4 is used as the
     new reference slice.

     The above algorithm is hidden from the user and  is  applied
     automatically  after the user selected ``Initialize'' in the
     ``Parameters'' menu. Please take a look of the skull xray to
     make  sure  the reference slice selected by the algorithm is
     inside the brain.

     If the reference slice selected by the program falls outside
     the brain, please select a slice inside the brain as the new
     reference slice in  the  ``Parameters''  menu.  Then  select
     ``Recompute''  to  recompute  all  the parameters. It is not
     necessary to define the reference points  (glabella,  inion,
     etc.) again.

     If you think the peak slice is fine,  You  can  either  type
     "Cancel"  or  you can simply type "Save" on the menu bar and
     then "Quit".


PRINTING
     Either the skull xray or the PET image can  be  saved  as  a
     window  dump  file  and  be  printed.  Right  now, it can be
     printed on the QMS color PostScript printer.  Do  not  print
     the  files  during  the peak hours; use batch files to print
     images in the evenings.



Neuro PET Group      Last change: 30-Sep-98                     1






User Commands                                             XRAY(1)



     Printing on the HP Laserjet is being studied.


DESCRIPTION
     xray creates a SXR file from an xray tracing and a PET image
     file.  The SXR file is used in _h_d_6_m_e_a_n and _t_8_8_m_e_a_n. The xray
     tracing should be a grayscale Sun raster file. The PET image
     can be a 7-slice PETT VI image or a 31-slice ECAT image. For
     ECAT images, it is recommended that the images  be  filtered
     by the Gaussian filter with half frequency 0.2.

     For parameters not specified, the following default  parame-
     ters are used:

      sxrfile:         .sxr is automatically used as the root.
      rasfile:         .ras is added to output file root to
                       search for Sun raster files.
      imgfile:         .img is added to output file root for
                       searching PET images.
      threshold:       40 for PETT VI emission scans; 50 for ECAT
                       emission scans; 80 for transmission scans.
      peaktype:        peak per scan.
      trim:            2 for transmission images; 0 for emission
                       images.

     The Sun raster file corresponding to  the  xray  tracing  is
     created  using  the  Microtech  scanner and a Macintosh com-
     puter. It is transferred to the Sun workstations where it is
     converted  to grayscale Sun raster format (See below: ``Pro-
     cessing Steps'').


PARAMETERS
      _s_x_r_f_i_l_e:       output file containing skull xray and PET
                     image parameters.

      _r_a_s_f_i_l_e:       xray raster image file.

      _i_m_g_f_i_l_e:       PET image file.

      _t_h_r_e_s_h_o_l_d:     percentage of threshold applied to PET
                     image to define the edges of the brain.

      _p_e_a_k_t_y_p_e:      1: threshold computed using peak for each
                     slice; 2: threshold computed using peak for
                     the whole scan.

      _t_r_i_m:          trimming factor.


PROCESSING STEPS
     Fran Miezin summaries the converting of skull xray  tracings



Neuro PET Group      Last change: 30-Sep-98                     2






User Commands                                             XRAY(1)



     to SXR files into a 4 step process:

        1. scan the tracing into the Macintosh using the
           Microtech flat bed scanner
        2. transfer the tracing pict file to the Sun workstations
        3. convert the pict file to a Sun Rasterfile (.ras) with
           appropriate orientation
        4. run the program xray (1) using the .ras  file  and  an
     image
           file and create an output .sxr file

     _S_c_a_n_n_i_n_g _i_n _t_h_e _S_k_u_l_l _T_r_a_c_i_n_g

     Place xray tracing on the Microtech  scanner  bed  with  the
     front  of  the  transparency against the glass, the glabella
     near the top and the inion near the bottom. Use the scale on
     the  scanner  bed to note the appropriate size of a bounding
     box which includes all of  the  tracing,  the  glabella  and
     inion  markers,  the ear marks and center slice marks. It is
     preferable but not necessary for this box to be close to the
     brain  dimension  as  opposed  to  being  much larger. Place
     several sheets of legal size white paper over the tracing to
     provide a uniformly white background for the tracing.

     Close the lid on the scanner.

     Select _M_i_c_r_o_t_e_k _C_o_l_o_r from the Apple menu, i.e., the icon in
     the  far  upper  left  corner  of the Macintosh screen.  The
     parameters in the window MUST be:
        SCSI 6
        Brightness:  Red:+21  Green:+21  Blue:+21
        Resolution:  100 dpi
        Scaling:     100%
        Save Format: PICT2 (Paint)

     Position the bounding box in the display window on the right
     of  the  scanner  bed to agree with the size that you deter-
     mined above. Press the SCAN button. For the filename use the
     subject  number followed by .pic. When scanning is finished,
     you can quit the application.

     _T_r_a_n_s_f_e_r_r_i_n_g ._p_i_c _F_i_l_e _t_o _S_u_n

     Select NCSA Telnet from the menu displayed  when  the  OnCue
     icon in the far upper right hand corner is pressed.

     Select Open Connection from the File menu.

     Enter petsun14 followed by a carriage return in the  Session
     Name window.

     Login to petsun14 and cd to the appropriate directory.



Neuro PET Group      Last change: 30-Sep-98                     3






User Commands                                             XRAY(1)



     Select Send FTP command from the Network menu.

     Enter binary to set the mode to binary transfer.

     Select Set Transfer Directory from the File menu and move to
     the  directory  which  contains  the .pic file. Click on the
     .pic filename and then click on Set Directory.

     Type get filename.pic.

     Type quit if successfully transferred.

     Type exit.

     _C_o_n_v_e_r_t_i_n_g ._p_i_c _F_i_l_e _t_o ._r_a_s _F_i_l_e

     cd to the directory on the Suns where the .sxr files  should
     be stored.

     Type picttoppm filename.pic | xv

     Using the right button on the mouse,  click  inside  of  the
     window to bring up the options window.

     Using one of the rotation arrows to rotate the image  either
     clockwise  or counter clockwise so that the image is upright
     in the display window.

     Check to make sure that all of the following are present  in
     the image:
        the complete tracing of the inner table of the skull
        the glabella and the inion points
        the mask for the anterior and posterior position of the
        central slice
        the ear markers locations

     Select ColEdit from the option window.

     Select gray from the left panel in the ColEdit window.

     Select save from the options menu. Save  display  as  a  Sun
     Rasterfile & grayscale image.

     Quit the program by clicking on the Quit button.

     _C_r_e_a_t_i_n_g _t_h_e ._s_x_r _F_i_l_e

     Run on the workstations. The program requires an image  file
     in  addition  to  the  .ras file.  This image file should be
     Gaussian filtered at 0.2 and from a scan with minimal  brain
     activation.




Neuro PET Group      Last change: 30-Sep-98                     4






User Commands                                             XRAY(1)



     Under the Parameters window enter the following values:
        ECAT scanner bed horizontal position during the xray
        ECAT scanner bed horizontal position during the scans
        The OFFSET value will  be  automatically  calculated  for
     ECAT
        studies.

     Select Xray and follow the  instructions.  After  the  skull
     xray  with  the AC-PC line is drawn on the screen, check the
     position of the reference slice. It  should  be  inside  the
     brain.  If  the reference slice is outside the brain, please
     go back to the ``Parameters'' menu and redefine  the  refer-
     ence slice to recompute the parameters.

     Select WinDump and save the drawing on the screen in a  win-
     dow dump file.

     Select PET and make sure that the template displayed  around
     each  of  the slices is correct; specifically look for noise
     pixels which make the template have an incorrect odd shape.

     Select Save.

     Select Quit.


HISTORY
     For xray version 3.0 by Tom Yang, the file type is 20.
     For xray version 3.1 by Tom Yang, the file type is 21.

     xray version 3.2 by Tom Yang (09/21/94 and 11/01/94)
      a) Allow the user to see which slice has the peak PET activity;
      b) Allow the user to select which slice to use for the threshold peak;
      c) Save this peak slice number in the sxr file;
      d) The file type is 22.

     The  following  calibrations  apply  to  the  953B  for  all
     acquisitions after Oct 1994:
      e) Distance from line on cassette to the center of slice 16 = 351.7 mm;
      f) Magnification factor at center of head = 0.9325.


NOTES
     If processing ECAT Version 7 images, scale them by  10000000
     or they won't have proper values in xray (i.e., their values
     will all be < 1).

     The raster file representing the x-ray is  assumed  to  have
     100 dots per inch.






Neuro PET Group      Last change: 30-Sep-98                     5






User Commands                                             XRAY(1)



EXAMPLES
     xray p2000
     xray p2000 p2000.ras
     xray p2000 p2000 p2000ho1
     xray p2000.sxr p2000.ras p2000ho1.img
     xray p2000.sxr p2000.ras p2000ho1.img 80
     xray p2000.sxr p2000.ras p2000ho1.img 20 1
     xray p2000.sxr p2000.ras p2000ho1.img 80 2 1


SEE ALSO
     xscale(1), t88mean(1), hd6mean(1)


AUTHOR
     Tom Yang: 1993.







































Neuro PET Group      Last change: 30-Sep-98                     6





npg/src/Old/xray/yang/xray.c

/*$Id: xray.c,v 1.12 1995/11/27 20:40:49 ty7777 Exp $*/
/*$Log: xray.c,v $
 * Revision 1.12  1995/11/27  20:40:49  ty7777
 * Updated for new 953B offset.
 *
 * Revision 1.11  1995/06/27  16:44:24  ty7777
 * Added processing for HR961 data.
 *
 * Revision 1.10  1995/02/13  22:03:31  ty7777
 * Added sxr_root to Setparameters ().
 *
 * Revision 1.9  1994/12/22  20:09:42  ty7777
 * Updated to regenerate PETT6 .sxr files.
 *
 * Revision 1.8  1994/11/02  15:36:21  ty7777
 * Added processing for new magnification factor and new ecat offset.
 *
 * Revision 1.7  1994/09/28  16:42:28  ty7777
 * Initialization for xscale.
 *
 * Revision 1.6  1994/09/20  16:00:19  ty7777
 * Updated for processing peakslice.
 *
 * Revision 1.5  1994/09/02  19:01:25  ty7777
 * Updated to incorporated new algorithm for setting the default reference slice.
 *
 * Revision 1.4  1994/08/30  20:15:18  ty7777
 * Updated to X11R5.
 *
 * Revision 1.3  1993/12/20  21:26:21  ty7777
 * Added recompute to GRAPHICS_DATA.
 *
 * Revision 1.2  1993/11/02  15:59:08  ty7777
 * Added makere_color.
 *
 * Revision 1.1  1993/10/07  16:26:37  ty7777
 * Initial revision
 **/

#ifndef lint
static char sccsid [] = "@(#)xray.c	11/16/92  Copyright Neural Pet Group, Washington University 1992";
#endif

static char rcsid [] = "$Header: /home/petsun4/ty7777/src/xray/RCS/xray.c,v 1.12 1995/11/27 20:40:49 ty7777 Exp $";

#include "xray.h"

#define MAX_STRINGS  64
#define MAX_ARGS     64

#include "xray.widgets"
#include "xray.callbacks"

char *messages []	= {"       Message:", 
			   "       Message: Select Glabella", 
			   "       Message: Select Inion", 
			   "       Message: Select Left Reference Slice Defining Point", 
			   "       Message: Select Right Reference Slice Defining Point", 
			   "       Message: Select Left Ear Marker", 
			   "       Message: Select Right Ear Marker"};

char list [] = "*;*;*;*"; /* any frame plane gate bed */

PointerStatus	cPointerState	= NoState;

main (argc, argv)
	int	argc;
	char	**argv;
{
	Arg			arglist [MAX_ARGS];
	BOOLEAN			acquired_after_oct94;
	BOOLEAN			assigned_threshold;
	BOOLEAN			check_delta_z;
	BOOLEAN			encoded;
	BOOLEAN			hrscanner;
	BOOLEAN			incorrect_delta_z;
	Colormap		cmap;                  /* resource id of color map */
	Cursor			define_cursor;
	DEFINING_POINTS		defining_points;
	DISPLAY_PET_DATA	display_pet_data;
	Display			*display;	/* X display */
	DRAW_DATA		draw_data;
	DRAWING_POINTS		drawing_points;
	GC			gcCanvas;       /* gc for canvas */
	GC			xorgc;      	/* gc for inverting */
	DATA_FILE_SELECTION	selectFileData;
	GRAPHICS_DATA		graphics_data;
	PARAMETER_MENU		parameter_menu;
	PARAMETER_DATA		initialize_parameter_data;
	PARAMETER_DATA		recompute_parameter_data;
	PARAMETER_DATA		cancel_parameter_data;
	PEAK_SLICE_CHANGED_DATA	recompute_peak_slice_data;
	PEAK_SLICE_CHANGED_DATA	cancel_peak_slice_data;
	PRINT_DATA		print_data;
	QUIT_DATA		quit_data;
	SAVESXR_DATA		savesxr_data;
	SAVESXR_DATA		savesxr_data_check_no;
	SXR_OUTPUT		sxr_output;
	Window			window;         /* window to display in */
	Window			root_window;         /* window to display in */
	XColor			cursor_bcolor;		/* RGB spec; contains pixel on return */
	XColor			cursor_fcolor;		/* RGB spec; contains pixel on return */
	XColor			marker_color;		/* RGB spec; contains pixel on return */
	XColor			pet_color;		/* RGB spec; contains pixel on return */
	XColor			xray_color;		/* RGB spec; contains pixel on return */
	XmString		xm_messages [MESSAGE_NUM];
	XmString		xm_buffers [MAX_ARGS];
	XRAY_DATA		xray_data;
	XSetWindowAttributes	wattr;		/* window attribute buffer */
	char			*ras_name;
	char			peaktype;
	char			*petfile;
	char			petimagetype;
	char			postscript_file [MAXLINE];
	char			*sxr_name;
	char			*sxr_root;
	char			szBuffer [MAXLINE];
	char			threshold;
	char			trim;
	int			cols, rows;
	int			i;
	int			j;
	int			n;
	int			imgfd;
	int			pettnum;	
	int			screen;		/* X screen */
	int			size;
	short int		num_slices;
	short int		temp;
	short int		*xray_tracing;
	llist			petlist;

	sxr_name		= argv [1];
	ras_name		= argv [2];
	petfile			= argv [3];
	threshold		= atoi (argv [4]);
	peaktype		= atoi (argv [5]);
	trim			= atoi (argv [6]);
	sxr_root		= argv [7];
	assigned_threshold	= atoi (argv [8]);

	/*
	 * Initializations.
	 */
	petlist	= llinit ();

	for (i = 0; i < XtNumber (messages); i++)
	{
		xm_messages [i]	= XmStringCreate (messages [i], XmSTRING_DEFAULT_CHARSET);	
	}

	check_delta_z		= FALSE;
	incorrect_delta_z	= FALSE;

	/*
	 * Read in X ray tracing.
	 */
	temp	= 1;
	ReadRaster (ras_name, &cols, &rows, &temp, TRUE);

	size		= cols * rows;
	xray_tracing	= (short *) malloc (size * sizeof (short));

	ReadRaster (ras_name, &cols, &rows, xray_tracing, FALSE);

	/*
	 * Get image number and determine if image is emission or transmission.
	 */
	imgfd	= open (petfile, O_RDONLY, 0);
	if (imgfd == -1)
	{
		fprintf (stderr, "Can't open %s\n", petfile); 
		exit (FAIL);
	}

	pettnum	= getimage_type (imgfd, petfile, &encoded, &num_slices);

	if (pettnum == ERROR_SCANNER)
	{
		fprintf (stderr, "Unrecongnized scanner type\n"); 
		exit (FAIL);
	}

	close (imgfd);

	/*
	 * Get magnification factor based on patient ID.
	 */
	sxr_output.film_position	= 0.0;
	sxr_output.number_slices	= num_slices;
	sxr_output.peaktype		= peaktype;
	sxr_output.threshold		= threshold;
	sxr_output.peakslice		= PEAK_SLICE_UNDEFINED;
	sxr_output.xscale		= 0.0;
	acquired_after_oct94		= FALSE;
	hrscanner			= FALSE;

	/*
	 * Assign parameters that are constant.
	 */
	if (SetParameters (petfile, sxr_root, pettnum, encoded, &num_slices, &sxr_output,
	&petimagetype, &trim, &acquired_after_oct94, &hrscanner, assigned_threshold) == FAIL)
	{
		fprintf (stderr, "SetParameters Error.\n");
		exit (FAIL);
	}

	if (GetMagnificationFactor (sxr_root, acquired_after_oct94, 
	hrscanner, pettnum, &check_delta_z, &incorrect_delta_z, &sxr_output) == FAIL)
	{
		fprintf (stderr, "Error: in GetMagnificationFactor.\n");
		exit (FAIL);
	}

	/*
	 * Memory allocation.
	 */
	sxr_output.rlslicedim		= (float *) malloc (num_slices * sizeof (float));
	sxr_output.rlpetdim		= (float *) malloc (num_slices * sizeof (float));
	sxr_output.rlcenter		= (float *) malloc (num_slices * sizeof (float));
	sxr_output.vertical_petdim	= (float *) malloc (num_slices * sizeof (float));
	drawing_points.slices_ap	= (POINT2D *) malloc (num_slices * 2 * sizeof (POINT2D));

#include "xray.code"

	/*
	 * Fill the menu structrue.
	 */
	parameter_menu.frameParameters		= &frameParameters;
	parameter_menu.labelXRayFile		= &labelXRayFile;
	parameter_menu.textMagnificationFactor	= &textMagnificationFactor;
	parameter_menu.textFilmPosition		= &textFilmPosition;
	parameter_menu.textScanPosition		= &textScanPosition;
	parameter_menu.textOffset		= &textOffset;
	parameter_menu.labelPetFile		= &labelPetFile;
	parameter_menu.textPixelSize		= &textPixelSize;
	parameter_menu.textSliceSpacing		= &textSliceSpacing;
	parameter_menu.textReferenceSlice	= &textReferenceSlice;
	parameter_menu.textNumberOfSlices	= &textNumberOfSlices;
	parameter_menu.textThreshold		= &textThreshold;
	parameter_menu.petfile			= petfile;
	parameter_menu.petimagetype		= petimagetype;
	parameter_menu.trim			= trim;
	parameter_menu.pettnum			= pettnum;
	parameter_menu.encoded			= encoded;
	parameter_menu.ecat_offset		= sxr_output.scan_position;
	parameter_menu.sxr_output		= &sxr_output;
	parameter_menu.petlist			= petlist;

	XtAddCallback (cascadeParameter, XmNactivateCallback, ShowParmMenuCallback, &parameter_menu);

	XtRealizeWidget (appShell);

	/* 
	 * Get the current display and screen. 
	 */
	display		= XtDisplay (appShell);
	screen		= DefaultScreen (display);
	define_cursor	= XCreateFontCursor (display, XC_crosshair);

        /* 
	 * Get the drawing area window. 
	 */
	root_window	= XtWindow (appShell);
	window		= XtWindow (mainWindow_ww);

	/* 
	 * Set backing store on window. 
	 */
	wattr.backing_store	= WhenMapped;
	XChangeWindowAttributes (display, window, CWBackingStore, &wattr);

	/* 
	 * Set the graphics context. 
	 */
	gcCanvas	= XCreateGC (display, window, 0, 0);
	xorgc		= CreateXorGC (mainWindow_ww);

	/* 
	 * Define foreground color. 
	 */
	cmap			= DefaultColormap (display, DefaultScreen (display));

	xray_color.red		= 65535;
	xray_color.green	= 65535;
	xray_color.blue		= 65535;
	if (XAllocColor (display, cmap, &xray_color) == 0)
	{
		fprintf (stderr, "Cann't allocate color\n");
	} 
	XSetForeground (display, gcCanvas, xray_color.pixel);

	marker_color.red	= 65535;
	marker_color.green	= 0;
	marker_color.blue	= 0;
	if (XAllocColor (display, cmap, &marker_color) == 0)
	{
		fprintf (stderr, "Cann't allocate color\n");
	} 

	pet_color.red	= 0;
	pet_color.green	= 0;
	pet_color.blue	= 65535;
	if (XAllocColor (display, cmap, &pet_color) == 0)
	{
		fprintf (stderr, "Cann't allocate color\n");
	} 

	cursor_fcolor.red	= 0;
	cursor_fcolor.green	= 65535;
	cursor_fcolor.blue	= 0;
	if (XAllocColor (display, cmap, &cursor_fcolor) == 0)
	{
		fprintf (stderr, "Cann't allocate color\n");
	} 

	cursor_bcolor.red	= 0;
	cursor_bcolor.green	= 0;
	cursor_bcolor.blue	= 0;
	if (XAllocColor (display, cmap, &cursor_bcolor) == 0)
	{
		fprintf (stderr, "Cann't allocate color\n");
	} 

	XRecolorCursor (display, define_cursor, &cursor_fcolor, &cursor_bcolor);

	draw_data.display	= display;
	draw_data.window	= window;
	draw_data.gc		= gcCanvas;

	xray_data.draw_data	= &draw_data;
	xray_data.cursor	= define_cursor;
	xray_data.xray_color	= xray_color.pixel;
	xray_data.cols		= cols;
	xray_data.rows		= rows;
	xray_data.xray_tracing	= xray_tracing;
	xray_data.message	= cascadeMessage;
	xray_data.xm_message	= xm_messages;
	XtAddCallback (cascadeXRay, XmNactivateCallback, DisplayXRayCallback, &xray_data);

	/*
	strcpy (postscript_file, sxr_root);
	strcat (postscript_file, ".ps");
	print_data.draw_data		= &draw_data;
	print_data.cursor		= define_cursor;
	print_data.drawing_points	= &drawing_points;
	print_data.cols			= cols;
	print_data.rows			= rows;
	print_data.xray_tracing		= xray_tracing;
	print_data.postscript_file	= postscript_file;
	XtAddCallback (cascadeSaveWinDump, XmNactivateCallback, PrintCallback, &print_data);
	*/

	XtAddCallback (XmFileSelectionBoxGetChild (winDumpFileSelection, 
			XmDIALOG_HELP_BUTTON), XmNactivateCallback, 
			CancelCallback, winDumpFileSelection);

	selectFileData.draw_data	= &draw_data;
	selectFileData.dialog		= winDumpFileSelection;
	XtAddCallback (XmFileSelectionBoxGetChild (winDumpFileSelection, 
			XmDIALOG_CANCEL_BUTTON), XmNactivateCallback, 
			SaveWinDumpCallback, &selectFileData);

	display_pet_data.draw_data		= &draw_data;
	display_pet_data.pet_color		= pet_color.pixel;
	display_pet_data.peak_color		= marker_color.pixel;
	display_pet_data.xray_color		= xray_color.pixel;
	display_pet_data.pettnum		= pettnum;
	display_pet_data.sxr_output		= &sxr_output;
	display_pet_data.petlist		= petlist;
	display_pet_data.dialog			= &framePeakSlice;
	display_pet_data.text_peak_slice	= &textPeakSlice;
	XtAddCallback (cascadePet, XmNactivateCallback, DisplayPetCallback, &display_pet_data);

	savesxr_data.sxr_name		= sxr_name;
	savesxr_data.petfile		= petfile;
	savesxr_data.sxr_root		= sxr_root;
	savesxr_data.incorrect_delta_z	= incorrect_delta_z;
	savesxr_data.sxr_output		= &sxr_output;
	if (!check_delta_z)
	{
		XtAddCallback (cascadeSave, XmNactivateCallback, SaveSXRCallback, &savesxr_data);
	}
	else
	{
		XtAddCallback (cascadeSave, XmNactivateCallback, ShowMenuCallback, checkDeltazQuestion);

		savesxr_data.incorrect_delta_z	= TRUE;
		XtAddCallback (XmMessageBoxGetChild (checkDeltazQuestion, XmDIALOG_OK_BUTTON),
		XmNactivateCallback, SaveSXRCallback, &savesxr_data);

		savesxr_data_check_no.sxr_name		= sxr_name;
		savesxr_data_check_no.petfile		= petfile;
		savesxr_data_check_no.sxr_root		= sxr_root;
		savesxr_data_check_no.incorrect_delta_z	= FALSE;
		savesxr_data_check_no.sxr_output	= &sxr_output;
		XtAddCallback (XmMessageBoxGetChild (checkDeltazQuestion, XmDIALOG_CANCEL_BUTTON),
		XmNactivateCallback, SaveSXRCallback, &savesxr_data_check_no);
	}

	quit_data.xray_tracing		= xray_tracing;
	quit_data.rlslicedim		= sxr_output.rlslicedim;
	quit_data.rlpetdim		= sxr_output.rlpetdim;
	quit_data.rlcenter		= sxr_output.rlcenter;
	quit_data.vertical_petdim	= sxr_output.vertical_petdim;
	quit_data.xm_message		= xm_messages;
	quit_data.petlist		= petlist;
	XtAddCallback (cascadeQuit, XmNactivateCallback, QuitCallback, &quit_data);

	/* 
	 * Add event handlers for mouse button events. 
	 */
	graphics_data.draw_data		= &draw_data;
	graphics_data.xorgc		= xorgc;
	graphics_data.defining_points	= &defining_points;
	graphics_data.drawing_points	= &drawing_points;
	graphics_data.cols		= cols;
	graphics_data.rows		= rows;
	graphics_data.xray_tracing	= xray_tracing;
	graphics_data.sxr_output	= &sxr_output;
	graphics_data.message		= cascadeMessage;
	graphics_data.recompute		= pushParametersRecompute;
	graphics_data.xm_message	= xm_messages;
	graphics_data.xray_color	= xray_color.pixel;
	graphics_data.marker_color	= marker_color.pixel;
	graphics_data.pet_color		= pet_color.pixel;

	initialize_parameter_data.node		= SELECT;
	initialize_parameter_data.parameter_menu	= &parameter_menu;
	initialize_parameter_data.graphics_data	= &graphics_data;
	XtAddCallback (pushParametersInitialize, XmNactivateCallback, 
			SelectParametersCallback, &initialize_parameter_data);

	recompute_parameter_data.node		= REDRAW;
	recompute_parameter_data.parameter_menu	= &parameter_menu;
	recompute_parameter_data.graphics_data	= &graphics_data;
	XtAddCallback (pushParametersRecompute, XmNactivateCallback, 
			SelectParametersCallback, &recompute_parameter_data);

	cancel_parameter_data.node		= CANCEL;
	cancel_parameter_data.parameter_menu	= &parameter_menu;
	cancel_parameter_data.graphics_data	= &graphics_data;
	XtAddCallback (pushParametersCancel, XmNactivateCallback, 
			SelectParametersCallback, &cancel_parameter_data);

	recompute_peak_slice_data.node			= REDRAW;
	recompute_peak_slice_data.parameter_menu	= &parameter_menu;
	recompute_peak_slice_data.display_pet_data	= &display_pet_data;
	XtAddCallback (pushPeakSliceRecompute, XmNactivateCallback, 
			PeakSliceChangedCallback, &recompute_peak_slice_data);

	cancel_peak_slice_data.node		= CANCEL;
	cancel_peak_slice_data.parameter_menu	= &parameter_menu;
	cancel_peak_slice_data.display_pet_data	= &display_pet_data;
	XtAddCallback (pushPeakSliceCancel, XmNactivateCallback, 
			PeakSliceChangedCallback, &cancel_peak_slice_data);

	XtAddEventHandler (mainWindow_ww, ButtonPressMask, FALSE, 
				ButtonDownEventHandler, &graphics_data);
	XtAddEventHandler (mainWindow_ww, ButtonMotionMask, FALSE, 
				ButtonMoveEventHandler, &graphics_data);
	XtAddEventHandler (mainWindow_ww, ButtonReleaseMask, FALSE, 
				ButtonReleaseEventHandler, &graphics_data);

	/* 
	 * Grab the mouse pointer on mouse events in drawing screen. 
	 */
	XGrabButton (XtDisplay (mainWindow_ww), AnyButton, AnyModifier, window, TRUE,
		ButtonPressMask | ButtonMotionMask | ButtonReleaseMask, GrabModeAsync, 
		GrabModeAsync, window, XCreateFontCursor (XtDisplay (mainWindow_ww), XC_fleur));
	XtMainLoop ();
}

